crates/edr_eth/src/signature/fakeable.rs

```diff
@@ -0,0 +1,217 @@
+                FakeableData::Fake { .. } => {
+                    unreachable!("fake signature must be initialized with address")
+                }
+            }
+        })
+    }
+}
+
+// We always assume that a decoded signature is a recoverable signature.
+impl<SignatureT> alloy_rlp::Decodable for Fakeable<SignatureT>
+where
+    SignatureT: alloy_rlp::Decodable + Recoverable + Signature,
+{
+    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
+        let signature = SignatureT::decode(buf)?;
+
+        Ok(Self {
+            data: FakeableData::Recoverable { signature },
+            address: OnceCell::new(),
+        })
+    }
+}
+
+// We need a custom implementation to avoid the struct being treated as an RLP
+// list.
+impl<SignatureT: alloy_rlp::Encodable + Recoverable + Signature> alloy_rlp::Encodable
+    for Fakeable<SignatureT>
+{
+    fn encode(&self, out: &mut dyn BufMut) {
+        match &self.data {
+            FakeableData::Fake { v } => {
+                if let Some(y_parity) = self.y_parity() {
+                    EcdsaWithYParity {
+                        r: self.r(),
+                        s: self.s(),
+                        y_parity,
+                    }
+                    .encode(out);
+                } else {
+                    let ecdsa = Ecdsa {
+                        r: self.r(),
+                        s: self.s(),
+                        v: *v,
+                    };
+
+                    println!("ecdsa: {ecdsa:?}");
+
+                    ecdsa.encode(out);
+                }
+            }
+            FakeableData::Recoverable { signature } => signature.encode(out),
+        }
+    }
+
+    fn length(&self) -> usize {
+        match &self.data {
+            FakeableData::Fake { v } => {
+                let v_length = if let Some(y_parity) = self.y_parity() {
+                    y_parity.length()
+                } else {
+                    v.length()
+                };
+
+                self.r().length() + self.s().length() + v_length
+            }
+            FakeableData::Recoverable { signature } => signature.length(),
+        }
+    }
+}
+
+impl<SignatureT: Recoverable + Signature> PartialEq for Fakeable<SignatureT> {
+    fn eq(&self, other: &Self) -> bool {
+        self.r() == other.r() && self.s() == other.s() && self.v() == other.v()
+    }
+}
+
+impl<SignatureT: Recoverable + Signature> Eq for Fakeable<SignatureT> {}
+
+#[cfg(feature = "serde")]
+impl<SignatureT: Recoverable + Signature> serde::Serialize for Fakeable<SignatureT> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        use serde::ser::SerializeMap;
+
+        let mut map = serializer.serialize_map(Some(3))?;
+        map.serialize_entry("r", &self.r())?;
+        map.serialize_entry("s", &self.s())?;
+        // We serialize the `v` field for backwards compatibility
```

Comment:
Could you expand on what needs to be backwards compatible?

Reply:
Addressed by https://github.com/NomicFoundation/edr/pull/498/commits/33d5df811d21864d06cb974daff96bc2d91c7e35